replica/table.cc

```diff
@@ -937,11 +947,27 @@ future<> tablet_storage_group_manager::maybe_split_compaction_group_of(size_t id
     return sg->split(split_compaction_options());
 }
 
+future<std::vector<sstables::shared_sstable>>
+tablet_storage_group_manager::maybe_split_sstable(const sstables::shared_sstable& sst) {
+    if (!tablet_map().needs_split()) {
+        co_return std::vector<sstables::shared_sstable>{sst};
+    }
+
+    auto& cg = compaction_group_for_sstable(sst);
+    auto holder = cg.async_gate().hold();
+    co_return co_await _t.get_compaction_manager().maybe_split_sstable(sst, cg.as_table_state(), split_compaction_options());
```

Comment:
How do we ensure that tablet_map doesn't change (split is not finalized) during split?

Reply:
split prepare -> repair -> split execute

repair coordinator holds erm, so under normal scenario this split happening in the context of repair cannot race with split finalization.

On restart, we'll split when adding sstable, before node is online.

If coordinator crashes, erm is released, and then we can have this split racing with finalization.

Theoretically we also have this race between repair and migration right? since there's a small window between sealing a sstable and adding it to the set. If coordinator crashes, the underlying tablet can be migrated away, before the sealed sstable is added to the set.

Do we need replica to also hold erm when adding sstable to the set?

Ah we do hold erm during repair writer, so I think we're safe.

```
void repair_writer_impl::create_writer(lw_shared_ptr<repair_writer> w) {
    if (_writer_done) {
        return;
    }
    replica::table& t = _db.local().find_column_family(_schema->id());
    rlogger.debug("repair_writer: keyspace={}, table={}, estimated_partitions={}", w->schema()->ks_name(), w->schema()->cf_name(), w->get_estimated_partitions());
    service::frozen_topology_guard topo_guard = service::null_topology_guard; // FIXME: propagate
    // The sharder is valid only when the erm is valid. Keep a reference of the erm to keep the sharder valid.
    auto erm = t.get_effective_replication_map();
    auto& sharder = erm->get_sharder(*(w->schema()));
    _writer_done = mutation_writer::distribute_reader_and_consume_on_shards(_schema, sharder, std::move(_queue_reader),
            streaming::make_streaming_consumer(sstables::repair_origin, _db, _view_builder, w->get_estimated_partitions(), _reason, is_offstrategy_supported(_reason), topo_guard),
    t.stream_in_progress()).then([w, erm] (uint64_t partitions) {
```

Do you want me to redundantly hold erm during split_sstable() to enforce the requirement? Or perhaps document it.

Reply:
@tgrabiec please let me know if my answer above satisfy you. like I explained, maybe_split_sstable() only happens in the context of repair / streaming, which holds the erm. on a hard reboot, we're able to continue the work before the sstable is loaded into the sstable set, and also before the node starts the messaging and group0 service.

Reply:
> Theoretically we also have this race between repair and migration right? since there's a small window between sealing a sstable and adding it to the set. If coordinator crashes, the underlying tablet can be migrated away, before the sealed sstable is added to the set.

This race is supposed to be fixed by using a `topology_guard` on the writer-side. We do use it, but it's armed only for streaming when doing a migration, for repair we pass a null guard around. 

> Ah we do hold erm during repair writer, so I think we're safe.

I think holding erm is not enough, since it can escape the barrier in split finalization:

1. handle_tablet_split_finalization() calls the barrier to wait for erm-holders. Barrier completes.
2. create_writer() takes erm, version=0, pre-split
4. handle_tablet_split_finalization() commits the split, all replicas change current erm to post-split
5. repair writer tries to add sstable, sees that need split is false, adds an sstable which crosses tablet boundary

We could protect against this by always checking if the sstable we try to add crosses the tablet boundary, and keep splitting it until it doesn't and atomically add it to the group. Then at least the invariant about non-overlapping would be satisfied based on local reasoning about the state of the local compaction groups, we wouldn't rely on any global reasoning about synchronization. This is more robust, because global synchronization is harder to prove.

Reply:
I think this can be addressed as a follow up so I queued the PR.

Reply:
"This race is supposed to be fixed by using a topology_guard on the writer-side. We do use it, but it's armed only for streaming when doing a migration, for repair we pass a null guard around." Why don't we do the same for repair? What's the downside?

Split compaction happens today with assumption of tablet count stability, it also spans a single tablet. If we go this direction, it might have to work with a snapshot. Will we have this problem you mentioned once we move repair to state machine?

Reply:
"2. create_writer() takes erm, version=0, pre-split"

but before step 2, won't we have repair coordinator waiting for coordinator to be quiesced? And if repair doesn't go through the coordinator, doesn't it mean it's also susceptible to other races for example with migration? since it can produce a sstable in a node to a tablet that was migrated away?

Reply:
we wouldn't silently add the sstable that overlaps 2 tablets, the same way we won't add a sstable silently if the tablet was migrated away. both scenarios result in a failure, and if that happen, it means a coordination side problem, right? for example, manager didn't disable balancing temporarily if it skipped repair coordinator somehow.

Reply:
>> "This race is supposed to be fixed by using a topology_guard on the writer-side. We do use it, but it's armed only for streaming when doing a migration, for repair we pass a null guard around." 
> Why don't we do the same for repair? What's the downside?

To arm topology_guard we need a valid session, which is currently propagated via group0, stored in tablet table. Moving repair to tablet transition state would naturally fit into this, otherwise we'd have to invent a new way to propagate and invalidate sessions, which will be difficult to do correctly without moving repair to tablet scheduler because when the repair coordinator dies, there is noone who would invalidate the session before erm is released. When repair is a transition, it will null the session field in system.tablet before transition ends, so before we move to a different transition on the tablet (including split finalization).   

> 
> Split compaction happens today with assumption of tablet count stability, it also spans a single tablet. If we go this direction, it might have to work with a snapshot. Will we have this problem you mentioned once we move repair to state machine?

No, if we move repair to state machine then we won't have this problem.

Reply:
> "2. create_writer() takes erm, version=0, pre-split"
> 
> but before step 2, won't we have repair coordinator waiting for coordinator to be quiesced? And if repair doesn't go through the coordinator, doesn't it mean it's also susceptible to other races for example with migration? since it can produce a sstable in a node to a tablet that was migrated away?

This scenario assumes the current implementation of repair. Yes, we're susceptible to other races. But there may be different severity level to them. If tablet was migrated away, the sstable will fail to attach because there is no compaction group. If it happens before cleanup, we don't care about losing the data because it belongs to a failed repair. Violating the invariant that sstables don't overlap tablet boundaries is more severe.